dex/pair/src/contexts/swap.rs

```diff
@@ -0,0 +1,300 @@
+    fn set_output_payments(&mut self, payments: ManagedVec<M, EsdtTokenPayment<M>>) {
+        self.output_payments = payments
+    }
+
+    fn get_output_payments(&self) -> &ManagedVec<M, EsdtTokenPayment<M>> {
+        &self.output_payments
+    }
+
+    fn get_opt_accept_funds_func(&self) -> &OptionalArg<ManagedBuffer<M>> {
+        &self.tx_input.args.opt_accept_funds_func
+    }
+
+    fn get_tx_input(&self) -> &dyn TxInput<M> {
+        &self.tx_input
+    }
+}
+
+impl<M: ManagedTypeApi> TxInputArgs<M> for SwapArgs<M> {
+    fn are_valid(&self) -> bool {
+        self.output_amount != 0 && self.output_token_id.is_esdt()
+    }
+}
+
+impl<M: ManagedTypeApi> TxInputPayments<M> for SwapPayments<M> {
+    fn are_valid(&self) -> bool {
+        self.input.amount != 0
+            && self.input.token_identifier.is_esdt()
+            && self.input.token_nonce == 0
+    }
+}
+
+impl<M: ManagedTypeApi> TxInput<M> for SwapTxInput<M> {
+    fn get_args(&self) -> &dyn TxInputArgs<M> {
+        &self.args
+    }
+
+    fn get_payments(&self) -> &dyn TxInputPayments<M> {
+        &self.payments
+    }
+
+    fn is_valid(&self) -> bool {
+        self.args.output_token_id != self.payments.input.token_identifier
+    }
+}
+
+impl<M: ManagedTypeApi> SwapContext<M> {
+    pub fn input_tokens_match_pool_tokens(&self) -> bool {
+        (self.tx_input.args.output_token_id == self.storage_cache.first_token_id
+            || self.tx_input.args.output_token_id == self.storage_cache.second_token_id)
+            && (self.tx_input.payments.input.token_identifier == self.storage_cache.first_token_id
+                || self.tx_input.payments.input.token_identifier
+                    == self.storage_cache.second_token_id)
+    }
+
+    pub fn get_payment(&self) -> &EsdtTokenPayment<M> {
+        &self.tx_input.payments.input
+    }
+
+    pub fn get_swap_args(&self) -> &SwapArgs<M> {
+        &self.tx_input.args
+    }
+
+    pub fn get_token_in(&self) -> &TokenIdentifier<M> {
+        &self.tx_input.payments.input.token_identifier
+    }
+
+    pub fn get_amount_in(&self) -> &BigUint<M> {
+        &self.tx_input.payments.input.amount
+    }
+
+    pub fn get_token_out(&self) -> &TokenIdentifier<M> {
+        &self.tx_input.args.output_token_id
+    }
+
+    pub fn get_amount_out_min(&self) -> &BigUint<M> {
+        self.get_amount_out()
+    }
+
+    pub fn get_amount_in_max(&self) -> &BigUint<M> {
+        self.get_amount_in()
+    }
+
+    pub fn get_amount_out(&self) -> &BigUint<M> {
+        &self.tx_input.args.output_amount
+    }
+
+    pub fn get_reserve_in(&self) -> &BigUint<M> {
+        let payment_token_id = &self.tx_input.payments.input.token_identifier;
+
+        if payment_token_id == &self.storage_cache.first_token_id {
+            &self.storage_cache.first_token_reserve
+        } else if payment_token_id == &self.storage_cache.second_token_id {
+            &self.storage_cache.second_token_reserve
+        } else {
+            unreachable!()
```

Comment:
Remove `unreachable!()`, as all it does it basically `panic!()` if that code section is ever reached. It doesn't actually remove the branch from the code, so you actually end up with even more code. If the code is unreachable anyway, simply remove the second `if` statement and return `&self.storage_cache.second_token_reserve` directly.

If you're unsure that's always the case for whatever reason, rewrite the method to either use `assert!()` or return an `Option<T>` instead. This applies to the other methods as well.

Reply:
unreachable -> assert is the intended behavior here